packages/ember-debug/lib/handlers.js

```diff
@@ -1,9 +1,23 @@
 import isPlainFunction from 'ember-debug/is-plain-function';
+import deprecate from 'ember-debug/deprecate';
 
 export let HANDLERS = { };
+export let testAsFunctionDeprecation = 'Functional arguments are ambiguous with constructors. ' +
+  'Please use !!Constructor for constructors, or an IIFE to compute the deprecation value. ' +
+  'In a future version of Ember functions will be treated as truthy values instead of being executed.';
 
 function normalizeTest(test) {
-  return isPlainFunction(test) ? test() : test;
+  if (isPlainFunction(test)) {
+    deprecate(
+      testAsFunctionDeprecation,
+      false,
+      { id: 'ember-debug.deprecate-test-as-function', until: '3.0.0' }
```

Comment:
Lets use `until: '2.5.0'` for this.  I am uncertain if this actually a public API (IMHO the current behavior of invoking constructors feels like a bug), but if we get this deprecated in 2.2 we will have a couple versions for things to get fixed...